dataplane/standalone/saiserver/switch.go

```diff
@@ -0,0 +1,165 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package saiserver
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/openconfig/lemming/dataplane/standalone/saiserver/attrmgr"
+
+	saipb "github.com/openconfig/lemming/dataplane/standalone/proto"
+)
+
+type saiSwitch struct {
+	saipb.UnimplementedSwitchServer
+	port   *port
+	vlan   *vlan
+	stp    *stp
+	vr     *virtualRouter
+	bridge *bridge
+	hostif *hostif
+	hash   *hash
+	mgr    *attrmgr.AttrMgr
+}
+
+func newSwitch(mgr *attrmgr.AttrMgr, s *grpc.Server) *saiSwitch {
+	sw := &saiSwitch{
+		port:   &port{},
+		vlan:   &vlan{},
+		stp:    &stp{},
+		vr:     &virtualRouter{},
+		bridge: &bridge{},
+		hostif: &hostif{},
+		hash:   &hash{},
+		mgr:    mgr,
+	}
+	saipb.RegisterSwitchServer(s, sw)
+	saipb.RegisterPortServer(s, sw.port)
+	saipb.RegisterVlanServer(s, sw.vlan)
+	saipb.RegisterStpServer(s, sw.stp)
+	saipb.RegisterVirtualRouterServer(s, sw.vr)
+	saipb.RegisterBridgeServer(s, sw.bridge)
+	saipb.RegisterHostifServer(s, sw.hostif)
+	saipb.RegisterHashServer(s, sw.hash)
+	return sw
+}
+
+// CreateSwitch a creates a new switch and populates its default values.
+func (sw *saiSwitch) CreateSwitch(ctx context.Context, _ *saipb.CreateSwitchRequest) (*saipb.CreateSwitchResponse, error) {
+	swID := sw.mgr.NextID()
+
+	// TODO: The port type is not a settable attribute, figure out a pattern for this.
+	cpuPortID := sw.mgr.NextID()
+	cpuPort := &saipb.PortAttribute{
+		Type: saipb.PortType_PORT_TYPE_CPU.Enum(),
+	}
+	sw.mgr.SetType(fmt.Sprint(cpuPortID), saipb.ObjectType_OBJECT_TYPE_PORT)
+	vlanResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.vlan.CreateVlan, &saipb.CreateVlanRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	stpResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.stp.CreateStp, &saipb.CreateStpRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	vrResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.vr.CreateVirtualRouter, &saipb.CreateVirtualRouterRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	brResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.bridge.CreateBridge, &saipb.CreateBridgeRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	trGroupResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.hostif.CreateHostifTrapGroup, &saipb.CreateHostifTrapGroupRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	hashResp, err := attrmgr.InvokeAndSave(ctx, sw.mgr, sw.hash.CreateHash, &saipb.CreateHashRequest{
+		Switch: swID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// These values are mostly meaningless, but clients expect these to be set.
```

Comment:
Please include some reference to where these magic numbers came from.

Reply:
done